wagtail/wagtailcore/models.py

```diff
@@ -1316,17 +1322,191 @@ def serve_password_required_response(self, request, form, action_url):
+            # The user is not permitted to see any pages, and therefore has no required ancestors, either.
+            pass
+
+        # Cache the paths for this user indefinitely. We'll clear this user's cached data manually when anything
+        # relevant changes, like changes to Group membership or permissions, or when a page is moved.
+        cache.set(cache_key, (permitted_paths, required_ancestors), None)
+
+    return permitted_paths, required_ancestors
+
+
+def convert_explorable_paths_to_Q(permitted_paths, required_ancestors):
+    """
+    Converts the output of get_explorable_page_paths() to a Q object which will filter a QuerySet down to the
+    appropriate set of Pages.
+    """
+    path_Qs = Q()
+    for path in permitted_paths:
+        path_Qs = path_Qs | Q(path__startswith=path)
+    for ancestor in required_ancestors:
+        path_Qs = path_Qs | Q(path=ancestor)
+    return path_Qs
+
+
+def clear_explorable_paths_cache(user=None):
+    """
+    Whenever a Page's path changes, or a Group's permissions change, or a User's Groups change, some or all of the
+    cached sets of explorable pages may now be invalid. Unfortunately, Django's caching system has very limited
+    ability to selectively clear cached data. So to avoid unnecessarily clearing unrelated cache data, we have to
+    be a bit brute forcey.
+
+    Pass in a user object to clear just that user's cached permitted path data. Otherwise, all cached permitted path
+    data will be cleared.
+    """
+    if user:
+        cache.delete(EXPLORABLE_PATHS_PREFIX.format(user.pk))
+    else:
+        # Delete any cached explorable paths that might exist for every user in the system.
+        cache.delete_many(
+            EXPLORABLE_PATHS_PREFIX.format(uid) for uid in get_user_model().objects.values_list('pk', flat=True)
+        )
+
+
+def get_closest_common_ancestor_path(user):
+    """
+    Returns the Closest Common Ancestor for the specified user, or None if the user has no explorable page paths.
+    """
+    permitted_paths, required_ancestors = get_explorable_page_paths(user)
+    if permitted_paths:
+        return required_ancestors[0]
+    else:
+        return None
+
+
+def get_navigation_menu_items(user):
     # Get all pages that appear in the navigation menu: ones which have children,
     # or are at the top-level (this rule required so that an empty site out-of-the-box has a working menu)
     pages = Page.objects.filter(Q(depth=2) | Q(numchild__gt=0)).order_by('path')
 
-    # Turn this into a tree structure:
+    # We need to treat the top of the user's explorable tree as if it were depth=2, or the depth_list algorithm won't
+    # work. So if a user isn't permitted to see pages at or below depth 2, we must change the apparent depth of each
+    # of their explorable pages.
+    depth_change = 0
+    if not user.is_superuser:
+        # For backward compatibility of the explorer_nav template tag, we accept "None" for the user argument, in which
```

Comment:
Great! I figured as much, but I wasn't sure if the tag was intentionally internal or not. I'll go ahead and remove that compatibility code.